File: clients/node/src/categories/executions.ts
import { IOneShotClient } from '../types/client.js';
import { TransactionExecution } from '../types/execution.js';
import { PagedResponse } from '../types/common.js';
import {
  transactionExecutionSchema,
  transactionExecutionListSchema,
  getTransactionExecutionSchema,
  listTransactionExecutionsSchema,
} from '../validation/execution.js';

export class Executions {
  constructor(private client: IOneShotClient) {}

  /**
   * Get a specific transaction execution
   * @param executionId The ID of the execution
   * @returns Promise<TransactionExecution>
   * @throws {ZodError} If the IDs are invalid
   */
  async get(executionId: string): Promise<TransactionExecution> {
    // Validate all parameters using the schema
    const validatedParams = getTransactionExecutionSchema.parse({
      executionId,
    });

    const response = await this.client.request<TransactionExecution>(
      'GET',
      `/executions/${validatedParams.executionId}`
    );

    // Validate the response
    return transactionExecutionSchema.parse(response);
  }

  /**
   * List transaction executions for a business
   * @param businessId The business ID to list executions for
   * @param params Optional filter parameters
   * @returns Promise<PagedResponse<TransactionExecution>>
   * @throws {ZodError} If the parameters are invalid
   */
  async list(
    businessId: string,
    params?: {
      pageSize?: number;
      page?: number;
      chainId?: number;
      status?: '0' | '1' | '2' | '3' | '4'; // Pending = 0, Submitted = 1, Completed = 2,	Retrying = 3,	Failed = 4,
      escrowWalletId?: string;
      transactionId?: string;
      apiCredentialId?: string;
      userId?: string;
    }
  ): Promise<PagedResponse<TransactionExecution>> {
    // Validate all parameters using the schema
    const validatedParams = listTransactionExecutionsSchema.parse({
      businessId,
      ...params,
    });

    const queryParams = new URLSearchParams();
    if (params) {
      Object.entries(params).forEach(([key, value]) => {
        if (value !== undefined) {
          queryParams.append(key, value.toString());
        }
      });
    }
    const queryString = queryParams.toString();
    const path = queryString
      ? `/business/${validatedParams.businessId}/transactions/executions?${queryString}`
      : `/business/${validatedParams.businessId}/transactions/executions`;

    const response = await this.client.request<PagedResponse<TransactionExecution>>('GET', path);

    // Validate the response
    return transactionExecutionListSchema.parse(response);
  }
}

File: clients/go/pkg/client/executions.go
package client

import (
	"context"

	swagger "github.com/UXlySoftware/1shot-clients/clients/go/internal/generated"
	"github.com/antihax/optional"
)

// Executions handles all transaction execution-related operations
type Executions struct {
	api        *swagger.TransactionExecutionApiService
	businessId string
}

// Get gets a specific transaction execution by its ID
func (e *Executions) Get(ctx context.Context, executionId string) (*swagger.TransactionExecution, error) {
	resp, _, err := e.api.ExecutionsTransactionExecutionIdGet(ctx, executionId)
	if err != nil {
		return nil, err
	}
	return &resp, nil
}

// List lists transaction executions for a business with optional filters
func (e *Executions) List(
	ctx context.Context,
	pageSize, page *int32,
	chainId *swagger.EChain,
	status *int32,
	escrowWalletId, transactionId, apiCredentialId, userId *string,
) ([]swagger.TransactionExecution, error) {
	opts := &swagger.TransactionExecutionApiBusinessBusinessIdTransactionsExecutionsGetOpts{}
	if pageSize != nil {
		opts.PageSize = optional.NewInterface(*pageSize)
	}
	if page != nil {
		opts.Page = optional.NewInterface(*page)
	}
	if chainId != nil {
		opts.ChainId = optional.NewInterface(*chainId)
	}
	if status != nil {
		opts.Status = optional.NewInt32(*status)
	}
	if escrowWalletId != nil {
		opts.EscrowWalletId = optional.NewInterface(*escrowWalletId)
	}
	if transactionId != nil {
		opts.TransactionId = optional.NewInterface(*transactionId)
	}
	if apiCredentialId != nil {
		opts.ApiCredentialId = optional.NewInterface(*apiCredentialId)
	}
	if userId != nil {
		opts.UserId = optional.NewInterface(*userId)
	}

	resp, _, err := e.api.BusinessBusinessIdTransactionsExecutionsGet(ctx, e.businessId, opts)
	if err != nil {
		return nil, err
	}
	return resp.Response, nil
}
